added telegram repository

Mansonita opened the pull request.

Commit:
added telegram repository

Commit:
added method GetProductsByBrandAndPrice and search by filters  to TelegramRepository

kapitan123 commented on YourEasyRent/DataBase/TelegramRepository.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace YourEasyRent.DataBase
+{
+    public class TelegramRepository : ITelegramRepository
```
It's still a ProductRepository. They are named after data that they return, in this case - Products.

kapitan123 commented on YourEasyRent/DataBase/TelegramRepository.cs:
```diff
@@ -0,0 +1,31 @@
+    public class TelegramRepository : ITelegramRepository
+    {
+
+        private readonly IMongoCollection<Product> _product;  // объявляет приватное поле _product в текущем классе TelegramRepository, которое будет использоваться для внедрения зависимости ITelegramRepository. Это поле будет доступно для использования в методах и конструкторах класса
```
It's not a product, it's a `productsCollection`

kapitan123 commented on YourEasyRent/DataBase/TelegramRepository.cs:
```diff
@@ -0,0 +1,31 @@
+        }
+        public async Task<IEnumerable<Product>> GetProductsByBrandAndPrice(string brand, decimal price)
+        {
+            var filter = Builders<Product>.Filter.And
```
Nice job, but there is easier way to do the same, you ca see it in my PR, but my way does not work for every case.

kapitan123 commented on YourEasyRent/DataBase/TelegramRepository.cs:
```diff
@@ -0,0 +1,31 @@
+        {
+            var filter = Builders<Product>.Filter.And
+                (Builders<Product>.Filter.Eq(_ => _.Brand, brand),
+                 Builders<Product>.Filter.Gte(_ => _.Price, price),  
```
This will never work. This is the same as:
`if( product.Price > 100 || product.Price < 100)`

kapitan123 commented on YourEasyRent/DataBase/Interfaces/ITelegramRepository.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface ITelegramRepository
+    {
+        Task<IEnumerable<Product?>> GetProductsByBrandAndPrice(string brand, decimal price);
```
It should never be null, so it should be `Task<IEnumerable<Product>>`

kapitan123 reviewed (commented):
Kind of on the right path

Commit:
deleted ITelegramRepository and TelegramRepository

Commit:
added files TelegramActionsHandler , ITelegramActionsHandler,  allocated a new class TelegramMenu

Commit:
started doing the class  TelegramPoller,
wrote comments on ptoductRepository class

Commit:
added telegramPoller

Commit:
added new node for search to "category of product"

Commit:
wrote  HandleUpdateAsync method,  added serviсes to programe file

Commit:
commented  the method with the price  to productRepository,  edited Program file

Commit:
corrected  _httpClient =  httpClient; to sephoraClient,
deleted cancellationToken

Commit:
created innerProductResponse  to get categoryNode

Commit:
added by  mothod MapNodesToProduct  checking for null, corrected priceNode for other html class

Commit:
added checking for null for  HtmlToProduct, added JsonProperty "classifications " to find category   for douglasClient

Commit:
added a  baselink to product url,
added new parameters of return to GetProductsByBrandAndCategory,
made swich/case for brand menu callBackQuery

Commit:
added State for bot,
changed script of bot menu

Commit:
divided the  productMessage into several parts,
added botState ReturnToMeinMenu by case

Commit:
replaced into update method arguments string brand, string name

Commit:
added worktime check for telegranbot, bit changу  menu

Commit:
made upsert method for productRepository and added it to listingController

Commit:
brushed up TelegramPoller

kapitan123 commented on YourEasyRent/Controllers/ListingController.cs:
```diff
@@ -38,25 +39,37 @@ public class ListingController : ControllerBase
+            var allListings = new List<Product>();
+            for (var pagenumber = 1; pagenumber < 4; pagenumber++)
+            {
+                var sephoralistings = await _sclient.FetchFromSephoraSection(section, pagenumber); //  сделать 2 listinga для сефоры и для дугласа, смерджить и результаты(list или join  погуглить как) и вернуть общий результат для обоих сайтов 
```
Check the implementation in my repo. In short you should have something like this

```
var clients = new List<IClient> {_sclient, _dclient};
clients.Select(x=> x.FetchFromSection(section, pagenumber).SelectMany(x=>x);
```

All your clients should implement the same interface, I'll also show you how to register them in such a way, so it is automatically put in List<IClient> .

Can be done in another PR.

kapitan123 commented on YourEasyRent/Controllers/ListingController.cs:
```diff
@@ -38,25 +39,37 @@ public class ListingController : ControllerBase
         }
 
+        [HttpPost("upsert")]
+        public async Task<IActionResult> UpsertProduct([FromBody] Product product)
```
It should be moved to ProductController. It should be [HttpPut] with no route. like 
`PUT api/products`

kapitan123 commented on YourEasyRent/DataBase/Interfaces/IProductRepository.cs:
```diff
@@ -5,16 +5,17 @@ namespace YourEasyRent.DataBase.Interfaces
+        Task<bool> Update(string brand, string name, Product updateproduct); 
+        Task<bool> Delete(string id);
+        Task<IEnumerable<Product>> GetProductsByBrandAndCategory(string brand, string category);
+        Task UpsertProduct(Product product);
```
You can also do `UpsertMany`

kapitan123 commented on YourEasyRent/DataBase/ProductRepository.cs:
```diff
@@ -7,77 +7,108 @@
-            var filter = Builders<Product>.Filter.Eq(_ => _.Id, id); // создаю фильтр для поиска продукта по его идентификатору.
-            var updateResult = await _product.ReplaceOneAsync(filter, updateProduct);// выполняет замену документа с учетом заданного фильтра.
+            var filter = Builders<Product>.Filter.And
+                (Builders<Product>.Filter.Eq(_ => _.Brand, brand),
```
Nice job, good enough. But I'll also show you how to do it, so it works really fast.

kapitan123 commented on YourEasyRent/DataBase/ProductRepository.cs:
```diff
@@ -7,77 +7,108 @@
+                (Builders<Product>.Filter.Eq(_ => _.Brand, product.Brand),
+                Builders<Product>.Filter.Eq(_ => _.Name, product.Name));
+
+            var update = Builders<Product>.Update
```
Shouldn't you set all the fields?

kapitan123 commented on YourEasyRent/Entities/BotState.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace YourEasyRent.Entities
+{
+    public enum BotState
```
Big brain move. Nice job.

kapitan123 commented on YourEasyRent/Services/ITelegramMenu.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public interface ITelegramMenu
+    {
+        InlineKeyboardMarkup meinMenu { get; }
```
Method names MUST start with a capital letter. E.G `MainMenu `

kapitan123 commented on YourEasyRent/Services/ITelegramMenu.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        InlineKeyboardMarkup meinMenu { get; }
+        InlineKeyboardMarkup brandMenu { get; }
+        InlineKeyboardMarkup categoryMenu { get; }
```
All of them are menus in a TelegramMenu class, so you can just name them => Main, Category, Brand

kapitan123 commented on YourEasyRent/Services/TelegramActionsHandler.cs:
```diff
@@ -0,0 +1,35 @@
+        public async Task<IEnumerable<string>> GetFilteredProductsMessage(string brand, string category)
+        {
+            var products = await _productRepository.GetProductsByBrandAndCategory(brand, category);
+            IEnumerable<string> productStrings = products.Select(p =>
```
```suggestion
            var productStrings = products.Select(p =>
```

kapitan123 reviewed (approved):
Nice job, a few minor comments, that ca nbc addressed in the next PR.
